combinetf2/fitter.py

```diff
@@ -0,0 +1,1417 @@
+    def loss_val(self):
+        val = self._compute_loss()
+        return val
+
+    @tf.function
+    def loss_val_grad(self):
+        with tf.GradientTape() as t:
+            val = self._compute_loss()
+        grad = t.gradient(val, self.x)
+
+        return val, grad
+
+    @tf.function
+    def loss_val_grad_hessp(self, p):
+        with tf.autodiff.ForwardAccumulator(self.x, p) as acc:
+            with tf.GradientTape() as grad_tape:
+                val = self._compute_loss()
+            grad = grad_tape.gradient(val, self.x)
+        hessp = acc.jvp(grad)
+
+        return val, grad, hessp
+
+    @tf.function
+    def loss_val_grad_hess(self, profile_grad=True):
+        with tf.GradientTape() as t2:
+            with tf.GradientTape() as t1:
+                val = self._compute_loss(profile_grad=profile_grad)
+            grad = t1.gradient(val, self.x)
+        hess = t2.jacobian(grad, self.x)
+
+        return val, grad, hess
+
+    def minimize(self):
+
+        def scipy_loss(xval):
+            self.x.assign(xval)
+            val, grad = self.loss_val_grad()
+            # print("scipy_loss", val)
+            return val.numpy(), grad.numpy()
+
+        def scipy_hessp(xval, pval):
+            self.x.assign(xval)
+            p = tf.convert_to_tensor(pval)
+            val, grad, hessp = self.loss_val_grad_hessp(p)
+            # print("scipy_hessp", val)
+            return hessp.numpy()
+
+        xval = self.x.numpy()
+
+        res = scipy.optimize.minimize(
+            scipy_loss, xval, method="trust-krylov", jac=True, hessp=scipy_hessp
+        )
+
+        xval = res["x"]
+
+        self.x.assign(xval)
+
+        print(res)
+
+        return res
+
+    def nll_scan(self, param, scan_range, scan_points, use_prefit=False):
+        # make a likelihood scan for a single parameter
+        # assuming the likelihood is minimized
+
+        idx = np.where(self.parms.astype(str) == param)[0][0]
+
+        # store current state of x temporarily
+        xval = tf.identity(self.x)
+
+        param_offsets = np.linspace(0, scan_range, scan_points // 2 + 1)
+        if not use_prefit:
+            param_offsets *= self.cov[idx, idx].numpy() ** 0.5
+
+        nscans = 2 * len(param_offsets) - 1
+        nlls = np.full(nscans, np.nan)
+        scan_vals = np.zeros(nscans)
+
+        # save delta nll w.r.t. global minimum
+        reduced_nll = self.reduced_nll().numpy()
+        # set central point
+        nlls[nscans // 2] = 0
+        scan_vals[nscans // 2] = xval[idx].numpy()
+        # scan positive side and negative side independently to profit from previous step
+        for sign in [-1, 1]:
+            param_scan_values = xval[idx].numpy() + sign * param_offsets
+            for i, ixval in enumerate(param_scan_values):
+                if i == 0:
+                    continue
+
+                self.x.assign(tf.tensor_scatter_nd_update(self.x, [[idx]], [ixval]))
+
+                def scipy_loss(xval):
```

Comment:
For the future: worth testing modifying the loss/grad/hess functions to imply 1 for the corresponding hessian element instead of 0, since this might allow the minimizer to converge more efficiently